internal/discovery/namespace_filter.go

```diff
@@ -0,0 +1,128 @@
+package discovery
+
+import (
+	"errors"
+
+	"github.com/newrelic/nri-kubernetes/v3/internal/config"
+	"github.com/newrelic/nri-kubernetes/v3/internal/storer"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	listersv1 "k8s.io/client-go/listers/core/v1"
+)
+
+// NamespaceFilterer provides an interface to filter namespaces.
+type NamespaceFilterer interface {
+	IsAllowed(namespace string) bool
+}
+
+// NamespaceFilter is a struct holding pointers to the config and the namespace lister.
+type NamespaceFilter struct {
+	c      *config.NamespaceSelector
+	lister listersv1.NamespaceLister
+	stopCh chan<- struct{}
+	storer storer.Storer
+	logger *log.Logger
+}
+
+// NewNamespaceFilter inits the namespace lister and returns a new NamespaceFilter.
+func NewNamespaceFilter(c *config.NamespaceSelector, client kubernetes.Interface, storer storer.Storer, logger *log.Logger, options ...informers.SharedInformerOption) *NamespaceFilter {
+	stopCh := make(chan struct{})
+
+	factory := informers.NewSharedInformerFactoryWithOptions(client, defaultResyncDuration, options...)
+
+	lister := factory.Core().V1().Namespaces().Lister()
+
+	factory.Start(stopCh)
+	factory.WaitForCacheSync(stopCh)
+
+	return &NamespaceFilter{
+		c:      c,
+		lister: lister,
+		stopCh: stopCh,
+		storer: storer,
+		logger: logger,
+	}
+}
+
+// IsAllowed checks given any namespace, if it's allowed to be scraped by using the NamespaceLister
+func (nf *NamespaceFilter) IsAllowed(namespace string) bool {
+	// By default, we scrape every namespace.
+	if nf.c == nil {
+		return true
+	}
```

Comment:
All these comments could be `log.Debugf` or `log.Tracef`
```suggestion
	if nf.c == nil {
		log.Tracef("Allowing %q as selector is nil", namespace)
		return true
	}
```

Reply:
Done in e57ba695a157903790cc9878ad0f6cbf032beb9f.